### DISCUSSION TITLE
Setting CSP with Material-UI (SSR) and Nginx

### DISCUSSION THREAD

**Discussion by Clumsy-Coder** [vercel/next.js#20716] [Help] [answered]
I have NextJS app that uses Material-UI rendered in Server Side. The NextJS app is hosted as a reverse proxy using Nginx.

I'm having issues allowing CSS in JS regarding CSP rules.

I'm aware you can use `unsafe-inline` for `style-src`. But I rather stick to best security practice.
I also checked using `nonce` but, I don't know how to inform Nginx about this.

Is there a way to solve this.

NOTE: 
- followed how to setup SSR for Material-UI
  - https://material-ui.com/styles/advanced/#next-js
  - https://github.com/mui-org/material-ui/tree/master/examples/nextjs

NOTE: 
- Why use Nginx as a server?
  - Nginx handles multiple requests efficiently. Hides NextJS server (only accessible through reverse proxy).

I get the following error in console
![image](https://user-images.githubusercontent.com/19594044/103484649-bb2e8080-4dad-11eb-8303-75f6e18684f9.png)

default.conf (nginx)
```nginx
# https://www.acunetix.com/blog/web-security-zone/hardening-nginx/

upstream nextjs_upstream {
  server localhost:3000;

  # We could add additional servers here for load-balancing
}

server {
  listen $PORT default_server;

  # redirect http to https. use only in production
  # if ($http_x_forwarded_proto != 'https') {
  #   rewrite ^(.*) https://$host$request_uri redirect;
  # }

  server_name _;

  server_tokens off;

  proxy_http_version 1.1;
  proxy_set_header Upgrade $http_upgrade;
  proxy_set_header Connection 'upgrade';
  proxy_set_header Host $host;
  proxy_cache_bypass $http_upgrade;

  # hide how is app powered. In this case hide NextJS is running behind the scenes.
  proxy_hide_header X-Powered-By;

  # set client request body buffer size to 1k. Usually 8k
  client_body_buffer_size 1k;
  client_header_buffer_size 1k;
  client_max_body_size 1k;
  large_client_header_buffers 2 1k;

  # ONLY respond to requests from HTTPS
  add_header Strict-Transport-Security "max-age=31536000; includeSubdomains; preload";

  # to prevent click-jacking
  add_header X-Frame-Options "DENY";

  # don't load scripts or CSS if their MIME type as indicated by the server is incorrect
  add_header X-Content-Type-Options nosniff;

  add_header 'Referrer-Policy' 'no-referrer';

  # Content Security Policy (CSP) and X-XSS-Protection (XSS)
  add_header Content-Security-Policy "
  default-src 'self'; 
  script-src 'self'; 
  object-src 'none'; 
  style-src 'self' https://fonts.googleapis.com ; 
  font-src fonts.gstatic.com ;
  form-action 'none'; 
  frame-ancestors 'none'; 
  base-uri 'none';" always;
  add_header X-XSS-Protection "1; mode=block";

  ssl_protocols TLSv1.2 TLSv1.3;
  ssl_prefer_server_ciphers on;

  location / {
    # limit request types to HTTP GET
    # ignore everything else
    limit_except GET { deny all; }

    proxy_pass http://nextjs_upstream;
  }
}
```

**Comment by leerob** [comment, marked as answer]
Hey folks, wanted to swing back here with an update. After digging through many different issues and discussions, I've [**made a new page in the documentation**](https://nextjs.org/docs/app/building-your-application/configuring/content-security-policy) ([PR](https://github.com/vercel/next.js/pull/54601)) specifically for Content Security Policy and nonces. This docs page: 

- Explains how to generate a `nonce` with Middleware
- Shows how to consume the `nonce` in a route with `headers()`
- Shows a complete CSP without needing to use any `unsafe`
- Shows how to ignore the `nonce` Middleware from running on prefetches / static assets

Further, we've patched some bugs and made improvements to `nonce` handling in Next.js that will be available in the latest `canary` version (for those of you time traveling from the future, upgrade to Next.js 13.5). We also updated the `with-strict-csp` example in the `examples/` folder, which is backlinked from the [new documentation page](https://nextjs.org/docs/app/building-your-application/configuring/content-security-policy).

Really hope this helps out, thank you all 🙏 I'll be closing this discussion out. To continue the discussion, [please go here](https://github.com/vercel/next.js/discussions/54907).
